test/integration/tracked_watermark_buffer.cc

```diff
@@ -94,5 +132,74 @@ std::pair<uint32_t, uint32_t> TrackedWatermarkBufferFactory::highWatermarkRange(
   return std::make_pair(min_watermark, max_watermark);
 }
 
+bool TrackedWatermarkBufferFactory::waitUntilTotalBufferedExceeds(
+    uint64_t byte_size, std::chrono::milliseconds timeout) {
+  absl::MutexLock lock(&mutex_);
+  auto predicate = [this, byte_size]() ABSL_SHARED_LOCKS_REQUIRED(mutex_) {
+    mutex_.AssertHeld();
+    return total_buffer_size_ >= byte_size;
+  };
+  return mutex_.AwaitWithTimeout(absl::Condition(&predicate), absl::Milliseconds(timeout.count()));
+}
+
+void TrackedWatermarkBufferFactory::removeDanglingAccounts() {
+  auto accounts_it = account_infos_.begin();
+  while (accounts_it != account_infos_.end()) {
+    auto next = std::next(accounts_it);
+
+    // Remove all "dangling" accounts.
+    if (accounts_it->first.use_count() == 1) {
+      account_infos_.erase(accounts_it);
+    }
+
+    accounts_it = next;
+  }
+}
+
+void TrackedWatermarkBufferFactory::setExpectedAccountBalance(uint64_t byte_size,
+                                                              uint32_t num_accounts) {
+  absl::MutexLock lock(&mutex_);
+  ASSERT(!expected_balances_.has_value());
+  expected_balances_.emplace(byte_size, num_accounts);
+}
+
+bool TrackedWatermarkBufferFactory::waitForExpectedAccountBalanceWithTimeout(
+    std::chrono::milliseconds timeout) {
+  return expected_balances_met_.WaitForNotificationWithTimeout(absl::FromChrono(timeout));
+}
+
+void TrackedWatermarkBufferFactory::checkIfExpectedBalancesMet() {
+  if (!expected_balances_ || expected_balances_met_.HasBeenNotified()) {
+    return;
+  }
+
+  removeDanglingAccounts();
+  if (account_infos_.size() < expected_balances_->num_accounts_) {
+    return;
+  }
+
+  // This is thread safe since this function should run on the only Envoy worker
+  // thread.
+  for (auto& acc : account_infos_) {
```

Comment:
The accounts may be owned by multiple dispatchers.  It is not safe to iterate through all accounts and ask for their balances; some of those accounts may be owned by dispatchers that are not the current one.